File: backend/app/models/audit.py
from datetime import datetime

from sqlalchemy import DateTime, Index, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class AuditLog(Base):
    __tablename__ = "audit_log"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    user_id: Mapped[int | None] = mapped_column(Integer, nullable=True)
    entity_type: Mapped[str] = mapped_column(String(64), nullable=False)
    entity_id: Mapped[int] = mapped_column(Integer, nullable=False)

    # 'CREATE','UPDATE','DELETE','APPROVE','REJECT'
    action: Mapped[str] = mapped_column(String(20), nullable=False)
    field_name: Mapped[str | None] = mapped_column(String(64), nullable=True)
    old_value: Mapped[str | None] = mapped_column(Text, nullable=True)
    new_value: Mapped[str | None] = mapped_column(Text, nullable=True)
    ip_address: Mapped[str | None] = mapped_column(String(45), nullable=True)
    occurred_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )

    __table_args__ = (
        Index("idx_audit_entity", "entity_type", "entity_id"),
        Index("idx_audit_user_date", "user_id", "occurred_at"),
    )

File: frontend/src/api/audit.ts
import { useQuery } from "@tanstack/react-query";

import { api } from "./client";

export interface AuditLogEntry {
  id: number;
  user_id: number | null;
  user_email: string | null;
  entity_type: string;
  entity_id: number;
  action: string;
  field_name: string | null;
  old_value: string | null;
  new_value: string | null;
  ip_address: string | null;
  occurred_at: string;
}

export interface PaginatedAuditLog {
  total: number;
  limit: number;
  offset: number;
  items: AuditLogEntry[];
}

export interface AuditLogParams {
  entity_type?: string;
  entity_id?: number;
  user_id?: number;
  action?: string;
  since?: string;
  until?: string;
  limit?: number;
  offset?: number;
}

export async function listAudit(params: AuditLogParams = {}): Promise<PaginatedAuditLog> {
  const { data } = await api.get<PaginatedAuditLog>("/audit/log", { params });
  return data;
}

export function useAuditLog(params: AuditLogParams) {
  return useQuery({
    queryKey: ["audit", "list", params],
    queryFn: () => listAudit(params),
    placeholderData: (prev) => prev,
  });
}

export const ENTITY_TYPES = [
  "portfolio_company",
  "portfolio_transaction",
  "valuation",
  "mis_submission",
] as const;

export const ACTIONS = [
  "CREATE",
  "UPDATE",
  "DELETE",
  "APPROVE",
  "REJECT",
  "UPLOAD",
  "MARK_CURRENT",
] as const;
